tests/conftest.py

```diff
@@ -0,0 +1,228 @@
+import threading
+import time
+import queue
+import sys
+import warnings
+import platform
+import itertools
+
+
+import pytest
+
+from PyQt5.QtWidgets import QApplication
+from PyQt5.QtCore import QTimer
+from _pytest.main import pytest_runtestloop as _pytest_runtestloop
+
+import ilastik.config
+
+from ilastik.utility.gui.threadRouter import ThreadRouter
+from ilastik.utility.itertools import pairwise
+from ilastik.shell.gui.startShellGui import launchShell
+
+
+GUI_TEST_TIMEOUT = 20  # Seconds
+
+
+def pytest_addoption(parser):
+    """Add command-line flags for pytest."""
+    parser.addoption("--run-legacy-gui", action="store_true",
+                     help="runs legacy gui tests")
+
+
+
+def pytest_pyfunc_call(pyfuncitem):
+    """
+    Defines protocol for legacy GUI test execution
+    """
+    if not is_gui_test(pyfuncitem):
+        return
+
+    bucket = [None]
+
+    def testfunc():
+        try:
+            return pyfuncitem.obj()
+        except Exception:
+            bucket[0] = sys.exc_info
+
+    thread = threading.Thread(target=testfunc)
+    thread.daemon = True
+
+    start = time.time()
+    thread.start()
+
+    while thread.is_alive():
+        time.sleep(0.2)
+        duration = time.time() - start
+        if duration > GUI_TEST_TIMEOUT:
+            raise TimeoutException()
+
+    exc_info = bucket[0]
+
+    if exc_info:
+        raise exc_info[1].with_traceback(exc_info[2])
+
+    return True
+
+
+@pytest.hookimpl(hookwrapper=True)
+def pytest_runtest_makereport(item, call):
+    """
+    Handle reports and chain fails for GUI tests
+    """
+    outcome = yield
+
+    if is_gui_test(item):
+        rep = outcome.get_result()
+
+        if rep.failed and call.excinfo is not None:
+            parent = item.parent
+            parent._failed = item
+            if call.excinfo.errisinstance(TimeoutException):
+                rep.wasxfail = "reason: Timeout"
+                rep.outcome = "skipped"
+
+
+def pytest_runtest_setup(item):
+    """
+    Handle chain fails for GUI tests
```

Comment:
it would be nice to add a few words here, what is so special about gui test chains, and why they need handling.

Reply:
Added clarification [comment](https://github.com/ilastik/ilastik/pull/1893/commits/ea14a2b0cc18532e9411c25373a73411f6ae8885)